bot/cogs/doc/cog.py

```diff
@@ -52,18 +53,83 @@
 NOT_FOUND_DELETE_DELAY = RedirectOutput.delete_delay
 
 
+class TODO_PLACEHOLDER:
+    __slots__ = ("queue", "item_events", "started", "fetch_lock")
+    results_dict: Dict[str, discord.Embed] = None
+
+    def __init__(self):
+        self.queue: List[DocItem] = list()
+        self.fetch_lock = asyncio.Lock()
+        self.started = False
+        self.item_events: Dict[DocItem, asyncio.Event] = {}
+
+    @classmethod
+    def set_results_dict(cls, results_dict: Dict[str, discord.Embed]):
+        """
+        Set dict for parse results for all instances.
+
+        Must be set before any parsing is attempted.
+        """
+        cls.results_dict = results_dict
+
+    def add_item(self, item):
+        """Add `item` into the parse queue."""
+        self.queue.append(item)
+
+    def put_to_front(self, item):
+        """Move `item` to the front of the parse queue."""
+        self.queue.remove(item)
+        self.queue.append(item)
+
+    async def wait_for_item(self, symbol_info, soup: Optional[BeautifulSoup] = None):
+        """
+        Block until `symbol_info` is parsed and assigned to `cls.results_dict`.
+
+        When not started, `soup` must be provided to start the parsing task;
+        after the initial request, providing the soup is unnecessary.
+        """
+        self.put_to_front(symbol_info)
+        item_event = asyncio.Event()
+        # First request
+        if not self.item_events:
+            log.info(f"Started {symbol_info.base_url + symbol_info.relative_url_path}")
+            self.item_events[symbol_info] = item_event
+            asyncio.create_task(self.parse_items(soup))
+        else:
+            self.item_events[symbol_info] = item_event
+        await item_event.wait()
+
+    async def parse_items(self, soup):
+        """Parse all item from `soup` and assign their result embeds into `cls.results_dict`."""
+        while self.queue:
+            item = self.queue.pop()
+            embed = discord.Embed(
+                title=discord.utils.escape_markdown(item.name),
+                url=item.url,
+                description=get_symbol_markdown(soup, item)
+            )
+            self.results_dict[item.name] = embed
+            event = self.item_events.get(item)
+            if event is not None:
+                event.set()
+            await asyncio.sleep(0.1)
```

Comment:
We need to give the event loop a bit of control so the bot can stay alive, `0` as the sleep period maxes out the core the process is running in. Locally 0.1 halved that utilization (with around a 3x increase in the total time to parse some larger pages) but I don't know how this would reflect in the production bot

Reply:
I have no idea. Seems kind of hacky but I've seen this sort of thing done elsewhere before.

Reply:
Another option would be to run this in an executor, but then we'd need to introduce some locks to prevent potential race conditions from the queue item movement. Giving the event loop control in this way works fine since the parsing doesn't take *that* long but I agree that it feels a bit hacky and the sleep period may influence things differently on different machines

Reply:
This is where my concurrency knowledge fails me. I thought the separate thread would still eat all the cycles. At what point would the context switch occur if one thread is giving no breathing room and is maxing out the core?

Reply:
The main thread that's running asyncio is doing lots of IO so that's getting it a lot of opportunities for the executors to acquire the GIL; for a full utilization in the executors, the first thread should send a forced request after a certain amount of time to the thread that's holding the GIL which has to release it.
The behaviour you're describing only happens when the thread is doing something during which python can't handle the GIL like huge calculations where time is spent on a single instruction.
As an example
```py
import threading


def cpu_and_memory_consumer():
    a = []
    while True:
        a.append(0)


def cpu_consumer():
    while True:
        pass


threads = [
    threading.Thread(target=cpu_consumer),
    threading.Thread(target=cpu_and_memory_consumer),
    #threading.Thread(target=lambda: 1234567890**1234567890),
]
for thread in threads:
    thread.start()
    print("started", thread)
```
Both of the functions take all that's available to them and shouldn't voluntarily release the GIL, but when you look at the memory usage it's clearly increasing even with the `cpu_consumer` started first (main thread and then the memory consumer both got the GIL to do their own thing), but then if we uncomment the computational thread everything will stop once it gets the GIL (No big memory usage from the appends and the started print never happens).

I'm of course not familiar with the devops of the bot and don't know if maxing out the available cpu is a good thing or not, in the threads without any sleep periods or through a `0` asyncio sleep